lib/use-suggestions.js

```diff
@@ -0,0 +1,104 @@
+import {
+	useCallback, useEffect, useMemo, useReducer
+} from 'haunted';
+import {
+	action, reduce
+} from '@neovici/cosmoz-utils/lib/reduce';
+import { useKeyboard } from './use-keyboard';
+
+const initial = {},
+	init = action('INIT', resolve => length => resolve(length)),
+	up = action('UP'),
+	down = action('DOWN'),
+	highlight = action('HIGHLIGHT', resolve => index => resolve(index)),
+	range = action(
+		'RANGE',
+		resolve => ({
+			firstVisible: first, lastVisible: last
+		}) =>
+			resolve({
+				first,
+				last
+			})
+	),
+	reducer = reduce(initial, handle => [
+		handle(init, (_, { payload: length }) => ({
+			length
+		})),
+		handle(down, ({
+			index: i = -1, ...state
+		}) => {
+			const {
+					length, range
+				} = state,
+				index = i < length - 1 ? i + 1 : 0;
+			return {
+				...state,
+				index,
+				scrollToIndex:
+					range && (index < range.first || index >= range.last)
+						? { index }
+						: undefined
+			};
+		}),
+		handle(up, ({
+			index: i = -1, ...state
+		}) => {
+			const {
+					length, range
+				} = state,
+				index = i > 0 ? i - 1 : length - 1;
+
+			return {
+				...state,
+				index,
+				scrollToIndex:
+					range && (index > range.last || index <= range.first)
+						? {
+							index,
+							position: 'end'
+						}
+						: undefined
+			};
+		}),
+		handle(highlight, (state, { payload: index }) => ({
+			...state,
+			index,
+			scrollToIndex: undefined
```

Comment:
What happens if we add a new action to handle and forget to add `scrollToIndex: undefined` in the new state?
Maybe the actions that depend on eachother like range and highlight should be handled by a reducer and the rest with useEffect?
At this point I think actually that we should keep the effects as clean as possible (even if it means we can't make optimizations in the reducer to avoid re-renders) to keep it as simple as possible, and switch to a reducer when it can be motivated by performance or complex aggregations.